### PR TITLE
Issue 48 dynamic error thresholding

### PR DESCRIPTION
**Pull request by itinawi**
References #48

### CODE REVIEW

File: mlblocks_primitives/mlprimitives.timeseries_errors.extract_anomalies.json [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    "modalities": ["timeseries"],
+    "primitive": "mlprimitives.timeseries_errors.extract_anomalies",
+    "produce": {
+        "args": [
+            {
+                "name": "y_true",
+                "type": "np.array"
+            },
+            {
+                "name": "smoothed_errors",
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries_errors.extract_anomalies.json]: This should be a hyperparameter

File: mlblocks_primitives/mlprimitives.timeseries_errors.extract_anomalies.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                "type": "list"
+            },
+            {
+                "name": "window_size",
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries_errors.extract_anomalies.json]: Please remove any hyperparameters from here. This list should contain only the function arguments that are required during fit and predict, not the hyperparameters.

File: mlblocks_primitives/mlprimitives.timeseries_errors.extract_anomalies.json [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            },
+            {
+                "name": "anomalies_scores",
+                "type": "list"
+            }
+        ]
+    },
+    "hyperparameters": {
+        "fixed": {
+            "window_size": {
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries_errors.extract_anomalies.json]: I think that all these hyperparameters should be set as tunable.

File: mlblocks_primitives/mlprimitives.timeseries_errors.get_forecast_error.json [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        "output": [
+            {
+                "name": "moving_avg",
+                "type": "list"
+            }
+        ]
+    },
+    "hyperparameters": {
+        "fixed": {
+            "window_size": {
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries_errors.get_forecast_error.json]: All these should be set as tunable hyperparameters.

File: mlblocks_primitives/mlprimitives.timeseries_errors.get_forecast_error.json [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                "type": "int",
+                "default": 30
+            },
+            "smoothing_percent": {
+                "type": "float",
+                "default": 0.05
+            },
+            "smoothed": {
+                "type": "bool",
+                "default": True
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries_errors.get_forecast_error.json]: Make sure to write a valid JSON format. This should be `true` (lowercase)

File: mlblocks_primitives/mlprimitives.timeseries_errors.get_forecast_error.json [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            {
+                "name": "y_hat",
+                "type": "np.array"
+            },
+            {
+                "name": "y_true",
+                "type": "np.array"
+            },
+            {
+                "name": "window_size",
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries_errors.get_forecast_error.json]: Please remove any hyperparameters from this list.

File: mlprimitives/timeseries_errors.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        return errors
+
+    historical_error_window = window_size * batch_size * smoothing_percent
+    moving_avg = []
+    for i in range(len(errors)):
+        left_window = i - historical_error_window
+        right_window = i + historical_error_window + 1
+        if left_window < 0:
+            left_window = 0
+        if right_window > len(errors):
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: Please add a blank line above this one (to separate unrelated if statements)

File: mlprimitives/timeseries_errors.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            right_window = len(errors)
+        moving_avg.append(np.mean(errors[left_window:right_window]))
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: Please add a blank line above this one

File: mlprimitives/timeseries_errors.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+            y_true ():
+            smoothed_errors ():
+            window_size (int):
+            batch_size (int):
+            error_buffer (int):
+        Returns:
+    """
+    if len(y_true) <= batch_size * window_size:
+        raise ValueError("Window size (%s) larger than y_true (len=%s)." % (batch_size, len(y_true)))
+    num_windows = (len(y_true) - (batch_size * window_size)) / batch_size
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: Please add a blank line above this one.

File: mlprimitives/timeseries_errors.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+
+        if i == num_windows + 1:
+            curr_index = len(y_true)
+
+        window_smoothed_errors = smoothed_errors[prev_index:curr_index]
+        window_y_true = y_true[prev_index:curr_index]
+
+        epsilon, sd_threshold = compute_threshold(window_smoothed_errors, error_buffer)
+
+        window_anom_indices = get_anomalies(window_smoothed_errors, window_y_true, sd_threshold, i, anomalies_indices, error_buffer)
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: This line is too long. Please write it as
```
window_anom_indices = get_anomalies(
    window_smoothed_errors,
    window_y_true,
    sd_threshold,
    i,
    anomalies_indices,
    error_buffer
)
```

File: mlprimitives/timeseries_errors.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+    # group anomalies
+    anomalies_indices = sorted(list(set(anomalies_indices)))
+    anomalies_groups = [list(group) for group in mit.consecutive_groups(anomalies_indices)]
+    anomaly_sequences = [(g[0], g[-1]) for g in anomalies_groups if not g[0] == g[-1]]
+
+    # generate "scores" for anomalies based on the max distance from epsilon for each sequence
+    anomalies_scores = []
+    for e_seq in anomaly_sequences:
+        denominator = np.mean(smoothed_errors) + np.std(smoothed_errors)
+        score = max([abs(smoothed_errors[x] - epsilon) / denominator
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: Please write this as
```
score = max([
    abs(smoothed_errors[x] - epsilon) / denominator
    for x in range(e_seq[0], e_seq[1])
])

File: mlprimitives/timeseries_errors.py [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+                    for x in range(e_seq[0], e_seq[1])])
+        anomalies_scores.append(score)
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: Please add a blank line above this one.

File: mlprimitives/timeseries_errors.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+
+    return anomaly_sequences, anomalies_scores
+
+
+def compute_threshold(smoothed_errors, error_buffer, sd_limit=12.0):
+    """
+        Helper method for extract_anomalies().
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: Please fix the docstring format. It should be:

```
"""Helper method for the `extract_anomalies` method.

Calculates the epsilon (threshold) for anomalies.
```

Please note the formatting and the indentation.

File: mlprimitives/timeseries_errors.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+        below_epsilon, below_indices, above_epsilon = [], [], []
+
+        for i in range(len(smoothed_errors)):
+            e = smoothed_errors[i]
+            if e < epsilon:
+                # save to compute delta mean and delta std
+                # these are important for epsilon calculation
+                below_epsilon.append(e)
+                below_indices.append(i)
+            if e > epsilon:
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: Please add a blank line above this one.

File: mlprimitives/timeseries_errors.py [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+    accepted_range = upper_percentile - lower_percentile
+
+    anomaly_indices = []
+    max_error_below_e = 0
+
+    for i in range(len(smoothed_errors)):
+        if smoothed_errors[i] <= epsilon or smoothed_errors[i] <= 0.05 * accepted_range:
+            # not an anomaly
+            continue
+        for j in range(error_buffer):
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: Please add a blank line above this one.

File: mlprimitives/timeseries_errors.py [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+
+    # group anomalies into continuous sequences
+    anomaly_indices = sorted(list(set(anomaly_indices)))
+    groups = [list(group) for group in mit.consecutive_groups(anomaly_indices)]
+    E_seq = [(g[0], g[-1]) for g in groups if g[0] != g[-1]]
+
+    return E_seq, anomaly_indices, max_error_below_e
+
+
+def prune_anomalies(E_seq, smoothed_errors, max_error_below_e, anomaly_indices):
```
Comment (csala) [inline mlprimitives/timeseries_errors.py]: Do not use Upper case variable names. `X` should be the only exception.

File: mlblocks_primitives/mlprimitives.timeseries_errors.extract_anomalies.json (hunk 1 above)
Comment (itinawi) [reply mlblocks_primitives/mlprimitives.timeseries_errors.extract_anomalies.json]: Why should it be? It is the output of `get_forecast_error` once it takes in `y_true` and `y_hat` and contains the sum of squared errors. It's not something that we can tune.

**Pull request merged**